File: DPCodes/221-maximumSquare.cpp

#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    vector<vector<int>> sums_;
public:
  int maximalSquare(vector<vector<int>> &matrix) {
    int m = matrix.size();
    int n = matrix[0].size();
    sums_ = vector<vector<int>>(m + 1, vector<int>(n + 1, 0));

    for (int i = 1; i <= m; ++i) {
      for (int j = 1; j <= n; ++j) {
        sums_[i][j] = matrix[i - 1][j - 1] + sums_[i][j - 1] + sums_[i - 1][j] -
                      sums_[i - 1][j - 1];
      }
    }
    int ans = 0;
    for (int i = 1; i <= m; ++i) {
      for (int j = 1; j <= n; ++j) {
        for (int k = min(m - i + 1, n - j + 1); k > 0; --k) {
          int tmp = sums_[i + k - 1][j + k - 1] - sums_[i + k - 1][j - 1] -
                    sums_[i - 1][j + k - 1] + sums_[i - 1][j - 1];
          if (tmp == k * k) {
            ans = max(tmp, ans);
            break;
          }
        }
      }
    }
    return ans;
  }
};

int main() {
  vector<vector<int>> m = {
      //   {1, 0, 1},//, 0, 0},
      {1, 1, 1}, // 1, 1},
      {1, 0, 1}  // 1, 1},
                 //   {1, 0, 0, 1, 0}
  };
  cout << Solution().maximalSquare(m);
}

File: DPCodes/221-maximumSquare.py




def maxSquare(m):
    r=len(m)
    c=len(m[0])

    s = [[0 for _ in range(c+1)] for _ in range(r+1)]
    # s = [[0 for i in range(c+1)] for j in range(r+1)]

    for i in range(1,r+1):
        for j in range(1,c+1):
            s[i][j]=m[i-1][j-1]+s[i-1][j]+s[i][j-1]-s[i-1][j-1]

    ans=0
    for i in range(1,r+1):
        for j in range(1,c+1):
            for k in range(min(r-i+1,c-j+1),0,-1):
                tmp=s[i+k-1][j+k-1]-s[i+k-1][j-1]-s[i-1][j+k-1]+s[i-1][j-1]
                if tmp==k*k:
                    ans=max(ans,tmp)
    return ans


# m=[[1,0,1,0,0],
#    [1,0,1,1,1],
#    [1,1,1,1,1],
#    [1,0,0,1,0]]
m=[[1,1,1],
   [1,1,1]]
print(maxSquare(m))           